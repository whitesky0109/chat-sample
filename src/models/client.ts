import { UserPayload } from 'public/store/actions/userAction';
import { RoomPayload } from 'public/store/actions/roomAction';

export interface IMessage {
  roomId?: string;
  type: string;
  message: File | string;
}

export interface StoreState {
  user: UserPayload;
  room: RoomPayload;
}
