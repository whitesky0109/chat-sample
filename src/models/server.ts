export interface IMigration {
  migration_id: string;
  action: string;
  args: any[];
}

export interface IRepository {
  migrations(): IMigration[];
}

export interface IService {
  init(): Promise<any>;
}

export interface IMigrationLog {
  id: number;
  migration_id: string;
  action: string;
  success: boolean;
  error: string;
  timestamp: Date;
}

export interface IMigrationLogs {
  // tslint:disable-next-line: variable-name
  [migration_id: string]: IMigrationLog;
}

export interface IInvite {
  roomId: string;
  users: string[];
}

export interface IMessage {
  userId: string;
  type: string;
  message: ArrayBuffer | string;
  date: Date;
  isMy: boolean;
}

export interface IUser {
  [userId: string]: string | null;
}

export interface IRoomInfo {
  name: string;
  users: IUser;
  creator: string;
}

export interface IRoom {
  [roomId: string]: IRoomInfo;
}
