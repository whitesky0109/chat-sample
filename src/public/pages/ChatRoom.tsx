import React, { Component } from 'react';
import { RouteProps } from 'react-router';

// models
import { IMessage } from '../../models/client';
import { IMessage as IServerMessage } from '../../models/server';

// components
import InviteUser from '../Components/Invite';
import ChatHistory from '../Components/ChatHistory';

// etc
import { socket } from '../utils';
import ChatInput from '../Components/ChatInput';

export interface Props extends RouteProps {
  match: any;
}

export interface States {
  messages: IServerMessage[];
  isInvite: boolean;
}

export default class ChatRoom extends Component<Props, States> {
  roomId: string;
  state: States = {
    messages: [],
    isInvite: false,
  };

  componentWillMount() {
    const { match } = this.props;
    this.roomId = match ? match.params.roomId : null;
    socket.on('room/message', this.onMessage);
  }

  componentWillUnmount() {
    socket.off('room/message');
  }

  onMessage = (msg: IServerMessage) => {
    const { messages } = this.state;
    this.setState({
      messages: [...messages, msg],
    });
  }

  onSendMessage = (msg: IMessage) => {
    socket.emit('room/message', msg);
  }

  onInvite = () => {
    const { isInvite } = this.state;

    this.setState({ isInvite: !isInvite });
  }

  render() {
    const { messages, isInvite } = this.state;

    return <>
      {isInvite ? <InviteUser roomId={this.roomId} /> : null}
      <button onClick={this.onInvite} >inVite</button>

      <ChatHistory messages={messages} />
      <ChatInput roomId={this.roomId} onSendMessage={this.onSendMessage} />

    </>;
  }
}
