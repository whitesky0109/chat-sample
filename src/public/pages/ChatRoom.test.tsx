import React from 'react';
import { ReactWrapper, mount } from 'enzyme';
import { StoreState } from 'models/client';
import createMockStore from 'redux-mock-store';
import ChatRoom, { Props } from './ChatRoom';
import { Provider } from 'react-redux';

describe('ChatRoom', () => {
  let component: ReactWrapper;
  let chatRoom: ReactWrapper;

  const mock: any = jest.fn();
  const props: Props = {
    match: mock,
    location: mock,
    history: {
      goBack: mock,
    } as any,
  };

  const generateFakeStore = (initState: StoreState) => {
    const mockStore = createMockStore();
    return mockStore(initState);
  };

  it('render', () => {
    const store = generateFakeStore({
      user: {
        user: 'userid',
        users: {
          ['userid']: 'sockid1',
        },
      },
      room: {
        roomId: 'roomid',
        room: {
          ['roomid']: {
            creator: 'userid',
            name: 'roomname',
            users: {
              ['userid']: 'sockid1',
            },
          },
        },
      },
    });

    component = mount(<Provider store={store}>
      <ChatRoom {...props} />
    </Provider>);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('find ChatRoom component', () => {
    chatRoom = component.find('ChatRoom');
    expect(chatRoom).toHaveLength(1);
  });

});
