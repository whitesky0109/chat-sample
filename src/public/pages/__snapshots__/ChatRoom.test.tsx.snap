// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoom matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ChatRoom)
    history={
      Object {
        "goBack": [MockFunction],
      }
    }
    location={[MockFunction]}
    match={[MockFunction]}
  >
    <ChatRoom
      history={
        Object {
          "goBack": [MockFunction],
        }
      }
      location={[MockFunction]}
      match={[MockFunction]}
      room={
        Object {
          "roomid": Object {
            "creator": "userid",
            "name": "roomname",
            "users": Object {
              "userid": "sockid1",
            },
          },
        }
      }
      roomId="roomid"
      user="userid"
    >
      <div
        className="header"
      >
        <div
          className="title"
        >
          <div
            className="user-icon"
          >
            <i
              className="fa fa-users"
            />
          </div>
          <div
            className="about"
          >
            <div
              className="chat-with"
            >
              roomname
            </div>
            <div
              className="chat-num-messages"
            >
              created by 
              userid
            </div>
          </div>
        </div>
        <div
          className="btn-group"
        >
          <div>
            <div>
              사용자 수 : 
              1
            </div>
            <div>
              접속자 수 : 
              1
            </div>
          </div>
          <button
            className="btn"
            onClick={[Function]}
            title="사용자 초대"
          >
            <i
              className="fa fa-user"
            />
          </button>
          <button
            className="btn"
            onClick={[Function]}
            title="방 나가기"
          >
            <i
              className="fa fa-sign-out"
            />
          </button>
        </div>
      </div>
      <ChatHistory
        messages={Array []}
      >
        <div
          className="chat-history"
        >
          <ul />
          <div />
        </div>
      </ChatHistory>
      <ChatInput
        onSendMessage={[Function]}
      >
        <div
          className="chat-input clearfix"
        >
          <div
            className="file-input"
          >
            <label
              htmlFor="image"
            >
              <span
                aria-hidden="true"
                className="fa fa-upload"
              />
            </label>
            <input
              accept="image/*"
              className="image-input"
              id="image"
              onChange={[Function]}
              type="file"
            />
          </div>
          <input
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Type your message"
            type="text"
            value=""
          />
          <button
            onClick={[Function]}
          >
            Send
          </button>
        </div>
      </ChatInput>
    </ChatRoom>
  </Connect(ChatRoom)>
</Provider>
`;
